constraint-solver/src/solver/boolean_extractor.rs

```diff
@@ -1,71 +1,131 @@
+        // This is in turn equivalent to `right + z * c = 0`, where `z` is
+        // a new boolean variable.
+
+        // For example, if the constraint was `(2 * a + 2 * b) * (a + b + 10) = 0`, we would
+        // set `factor = 1 / 2`, such that `left * factor - right` is a constant.
+
+        // First, try to find a good factor so that `left` and `right`
+        // likely cancel out except for a constant. As a good guess,
+        // we try to match the coefficient of the first variable.
+        let factor = match (left.components().1.next(), right.components().1.next()) {
+            (Some((left_var, left_coeff)), Some((right_var, right_coeff)))
+                if left_var == right_var =>
+            {
+                right_coeff.field_div(left_coeff)
+            }
+            _ => T::one(),
+        };
+
+        // `constr = 0` is equivalent to `left * right = 0`
+        let offset = &(left.clone() * &factor) - right;
+        // We only do the transformation if `offset` is known, because
+        // otherwise the constraint stays quadratic.
+        let offset = offset.try_to_known()?;
+        // We know that `offset + right = left` and thus
+        // `constr = 0` is equivalent to `right * (right + offset) = 0`
+        // which is equivalent to `right + z * offset = 0` for a new
+        // boolean variable `z`.
+
+        if offset.is_zero() {
+            // In this special case, we do not need a new variable.
+            return Some(right.clone());
+        }
+
+        let key = -right * T::one().field_div(offset);
+        if self.substitutions.contains_key(&key) {
+            // We have already performed this transformation before.
+            return None;
+        }
+
+        if key.try_to_simple_unknown().is_some() {
+            // In this case we don't gain anything because the new variable `z` will just
+            // be equivalent to the single variable in `right`.
+            None
+        } else {
+            let z = var_dispenser();
+
+            self.substitutions.insert(key, z.clone());
+
+            // We return `right + z * offset == 0`, which is equivalent to the original constraint.
+            Some(right + &(GroupedExpression::from_unknown_variable(z) * offset))
+        }
+    }
+}
+
+impl<T: RuntimeConstant + Substitutable<V> + Hash, V: Clone + Eq + Ord + Hash>
+    BooleanExtractor<T, V>
+{
+    /// Applies the assignments to the stored substitutions.
+    pub fn apply_assignments(&mut self, assignments: &[VariableAssignment<T, V>]) {
+        if assignments.is_empty() {
+            return;
+        }
+        let (exprs, vars): (Vec<_>, Vec<_>) = self.substitutions.drain().unzip();
+        let exprs = apply_substitutions(
+            ConstraintSystem {
+                algebraic_constraints: exprs,
+                bus_interactions: vec![],
+            },
+            assignments.iter().cloned(),
+        )
+        .algebraic_constraints;
```

Comment:
Would be good to have `apply_substitutions` as a trait and implement it on `algebraic_constraint`, to avoid having to wrap things like this.

Reply:
Yes, it builds an index. I could implement a function next to `apply_substitutions` that takes a vec of exprs and a vec of substitutions.